Plugins/Mute/DataManager.cs

```diff
@@ -1,38 +1,119 @@
+                Expiration = DateTime.UtcNow,
+                AdminId = Utilities.IW4MAdminClient().ClientId,
+                MuteState = MuteState.Unmuted
+            };
+            client.SetAdditionalProperty(Plugin.MuteKey, clientMuteStateMeta);
+            return clientMuteStateMeta;
+        }
 
-        client.SetAdditionalProperty(Plugin.MuteKey, clientMuteState);
-        return clientMuteState;
+        // Populate meta so doesn't trigger infinite recursion in switch case.
+        // TODO: This sucks. I don't like it. It should be handled by MuteManager. Though, I don't know how.
+        clientMuteStateMeta = await WritePersistentData(client, Utilities.IW4MAdminClient(), DateTime.UtcNow,
+            MuteState.Unmuted, string.Empty);
+
+        switch (oldClientMuteState)
+        {
+            case MuteState.Muted:
+                // Run mute logic and update the return object.
+                await Plugin.MuteManager.Mute(client.CurrentServer, Utilities.IW4MAdminClient(), client,
+                    null, _translationLookup["PLUGINS_MUTE_MIGRATED"]);
+                clientMuteStateMeta = await ReadPersistentData(client);
+                break;
+            case MuteState.Unmuting:
+            case MuteState.Unmuted:
+            default:
+                break;
+        }
+
+        // Return null or the populated meta.
+        return clientMuteStateMeta;
     }
 
-    public async Task WritePersistentData(EFClient client, MuteState state)
+    public async Task<MuteStateMeta> WritePersistentData(EFClient client, EFClient adminId, DateTime? dateTime,
+        MuteState state, string reason)
     {
-        await _metaService.SetPersistentMeta(Plugin.MuteKey, state.ToString(), client.ClientId);
-        client.SetAdditionalProperty(Plugin.MuteKey, state);
+        var clientMuteStateMeta = client.GetAdditionalProperty<MuteStateMeta>(Plugin.MuteKey) ??
+                                  await _metaService.GetPersistentMetaValue<MuteStateMeta>(Plugin.MuteKey,
+                                      client.ClientId);
+
+        // Update existing meta
+        if (clientMuteStateMeta is not null)
+        {
+            clientMuteStateMeta.ClientId = client.ClientId;
+            clientMuteStateMeta.AdminId = adminId.ClientId;
+            clientMuteStateMeta.MuteState = state;
+            clientMuteStateMeta.Expiration = dateTime;
+            clientMuteStateMeta.Reason = reason;
+        }
+        else // Create new meta
+        {
+            clientMuteStateMeta = new MuteStateMeta
+            {
+                ClientId = client.ClientId,
+                Reason = reason,
+                Expiration = dateTime,
+                AdminId = adminId.ClientId,
+                MuteState = state
+            };
+        }
+
+        await _metaService.SetPersistentMetaValue(Plugin.MuteKey, clientMuteStateMeta, client.ClientId);
+        client.SetAdditionalProperty(Plugin.MuteKey, clientMuteStateMeta);
```

Comment:
Don't need to set this here because in theory the data passed in should match the additional property instance exactly.

If needed this logic should be in mute manager